### PR TITLE
Implemented time range querying

### PR DESCRIPTION
**Pull request by simonszu**
I have implemented time range querying. Unfortunately, Ruby's `Date::parse` method only accepts full dates, speak: with year, month and day, so it is not possible to parse incomplete dates which specify time ranges. 
To solve this, i have created some switches, combined with regular expressions and tried to parse the given date by hand. I do not know what formats `Date::parse` accepts, so my regexps only accept `YYYY-MM-DD` and `YYYY/MM/DD` for now. If desired, i can add other formats with even more regular expressions. 

For now, both acceptable formats are mentioned in the README, and the reference to the `Date::parse` method was removed from the README.

Furthermore, i fixed https://github.com/igneus/calendarium-romanum/issues/10. The solution you posted in this issue didn't work for me (at least with Ruby 2.4.2), so i solved it with a ternary operator. I hope it's okay. 

And last, but not least, a change so small that i didn't want to create another branch for it: I added built gems to the gitignore. I don't know your workflow, but i used `gem build` and `gem install` every time i wanted to test my code manually, which resulted in a built gem inside the work dir, which one could accidentally commit when using `git add .`, so i excluded it in the gitignore.

### CODE REVIEW

File: lib/calendarium-romanum/cli.rb [hunk 1]
```diff
@@ -17,29 +17,37 @@ def query(date_str=nil)
...
-        if date_str
-          begin
-            Date.parse(date_str)
-          rescue ArgumentError
-            die! 'Invalid date.'
+      if date_str
+        begin 
+          print_single_date(Date.parse(date_str), sanctorale)
+        rescue ArgumentError
+          if ((date_str =~ /((\d{4})-(\d{2})-(\d{2}))\z/) || (date_str =~ /((\d{4})\/(\d{2})\/(\d{2}))\z/)) # Year and month and day are given
```
Comment (igneus) [inline lib/calendarium-romanum/cli.rb]: There is a lot of duplicate code here. Please, extract the date parsing logic to a new class (`CalendariumRomanum::Util` namespace is intended for internal tools like this).

Possible bonus points for handling all these cases in a single regular expression. :smile:

File: lib/calendarium-romanum/cli.rb [hunk 2]
```diff
@@ -115,5 +123,26 @@ def die!(message, code=1)
       STDERR.puts message
       exit code
     end
+
+    private
+
+    def print_single_date(date, sanctorale)
+      calendar = Calendar.for_day(date, sanctorale)
```
Comment (igneus) [inline lib/calendarium-romanum/cli.rb]: It's inefficient to create a new `Calendar` instance for each day to print. Instead, use one `PerpetualCalendar` instance for all days.

**Comment by igneus** [comment]
Tip: when manually testing the executable, I don't build and install the gem (except when testing if the gem builds and installs correctly), but execute it in a more straightforward way: `ruby -I ./lib bin/calendariumrom ...`

File: lib/calendarium-romanum/cli.rb (hunk 1 above)
Comment (simonszu) [reply lib/calendarium-romanum/cli.rb]: No problem. Parser and even a single regular expression are already implemented. I will think about your second issue.

**Comment by simonszu** [comment]
So, as you can see, i have written a new class for parsing the date ranges, and discovered that i can use the other range classes specified in `CalendariumRomanum::Util`. This works for now. 

But now i am struggling with the calendar format. As you can see in my [WIP] commit, i try to use one PerpetualCalendar creating one Calendar for each time range, and providing this Calendar to the `print_single_date` method. However, there is the classical error i am making with the difference Liturgical year <=> Civil year. 

Can you give me a hint, please? 

Hm - or, wait. I didn't see the "if you need them" comment in the README-section for the PerpetualCalendar. I'll continue fiddling with this stuff...

**Comment by simonszu** [comment]
Well, i think i figured it out. My mistake was, that i created a `PerpetualCalendar` instance, and then a `Calendar` instance from it. Of course, this did not work. After deleting the creation of the `Calendar` instance, it worked. 

So, now we can debate if it is clean enough code when the parser either creates timerange-enumerator-objects, or just a single date, and the query command switching these objects afterwards - but the only workaround for this would be some kind of single-day enumerator object, which also responds to the `.each` command, but i think this is ugly as well.

**Comment by simonszu** [comment]
Well...seems, at least, i still do need some help with calendars. As you see, i have used `PerpetualCalendar`. Turns out, `PerpetualCalendar` does not return any celebrations. So, i need to create a new liturgical calendar object for each day? But then i am at the very beginning again, with creating lots of liturgical calendar objects, which is not desired. 

Where's my mistake? How would you solve it?

**Comment by igneus** [comment]
> but the only workaround for this would be some kind of single-day enumerator object, which also responds to the .each command, but i think this is ugly as well.

So do I. In this case I would not use `Util::Year` and `::Month`, but simply a `Range` of `Date`s. It is possible to make a `Range` of only one value, e.g. `1 .. 1`.

**Comment by igneus** [comment]
> Turns out, PerpetualCalendar does not return any celebrations.

It does return celebrations, but no *sanctorale* celebrations. The reason is that

```ruby
pcal = PerpetualCalendar.new
```

does not provide the newly created instance with any sanctorale data. Change to

```ruby
pcal = PerpetualCalendar.new sanctorale: sanctorale
```

**Comment by simonszu** [comment]
Okay, works now. I also have added the `coverage/` folder created by rspec to the gitignore.

File: lib/calendarium-romanum/cli.rb [hunk 3]
```diff
@@ -115,5 +107,25 @@ def die!(message, code=1)
       STDERR.puts message
       exit code
     end
+
+    private
+
+    def print_single_date(calendar, date, sanctorale)
```
Comment (igneus) [inline lib/calendarium-romanum/cli.rb]: argument `sanctorale` is unused, remove it

File: lib/calendarium-romanum/util.rb [hunk 4]
```diff
@@ -23,6 +23,7 @@ def enumeration_over?(date)
 
     # enumerates days of a year
     class Year < DateEnumerator
+      attr_reader :start
```
Comment (igneus) [inline lib/calendarium-romanum/util.rb]: It seems `Year#start` and `Month#start` are not used anywhere. No reason to add an attribute.

File: lib/calendarium-romanum/util.rb [hunk 5]
```diff
@@ -31,10 +32,31 @@ def initialize(year)
...
     class Month < DateEnumerator
+      attr_reader :start
       def initialize(year, month)
         @start = Date.new year, month, 1
         @prop = :month
       end
     end
+
+    class DateParser
+      attr_reader :payload
```
Comment (igneus) [inline lib/calendarium-romanum/util.rb]: Please rename `#payload` to something more meaningful. E.g. something like `#dates` or `#date_range` would do well.

File: lib/calendarium-romanum/cli.rb [hunk 6]
```diff
@@ -115,5 +107,25 @@ def die!(message, code=1)
       STDERR.puts message
       exit code
     end
+
+    private
```
Comment (igneus) [inline lib/calendarium-romanum/cli.rb]: No need for `private` here, we are already in a block of private methods. See a few lines above.

File: lib/calendarium-romanum/cli.rb [hunk 7]
```diff
@@ -17,29 +17,21 @@ def query(date_str=nil)
...
+
+      if date_str
+        begin
+          parsed_date = DateParser.new(date_str)
+          parsed_date.payload.each do |day|
+            print_single_date(pcal, day, sanctorale)
           end
-        else
-          Date.today
+        rescue
```
Comment (igneus) [inline lib/calendarium-romanum/cli.rb]: Unqualified `rescue` is bad practice. That reminds me of the fact that `DateParser` should have unit tests including error scenarios.

File: lib/calendarium-romanum/cli.rb (hunk 3 above)
Comment (simonszu) [reply lib/calendarium-romanum/cli.rb]: Done. Just posting it here as a checklist for myself.

File: lib/calendarium-romanum/util.rb (hunk 4 above)
Comment (simonszu) [reply lib/calendarium-romanum/util.rb]: Done. Just posting it here as a checklist for myself.

File: lib/calendarium-romanum/util.rb (hunk 5 above)
Comment (simonszu) [reply lib/calendarium-romanum/util.rb]: Done. It's called `#date_range` now

File: lib/calendarium-romanum/cli.rb (hunk 6 above)
Comment (simonszu) [reply lib/calendarium-romanum/cli.rb]: Done. Just posting it here as a checklist for myself.

File: lib/calendarium-romanum/cli.rb (hunk 7 above)
Comment (simonszu) [reply lib/calendarium-romanum/cli.rb]: Done. Created some tests for `DateParser` as well.

**Comment by igneus** [comment]
Thank you, code is ready to be merged now. I will ask you for a last single change: please, undelete `tmp/aruba/.keep` and delete `tmp/aruba/*.txt`. The files are deleted/created by aruba while executing tests and no changes in this directory should ever be committed.

It would also be useful to rename `spec/util_spec.rb` to `spec/date_parser_spec.rb`. (I plan to split files containing multiple classes to one-class-per-file in the near future.)

**Comment by simonszu** [comment]
No problem. I thought i already had re-added the keep-file, but as usual, `git add .` does not add hidden files ;)

**Comment by igneus** [comment]
Btw. I noticed you rebase and force-push (or do something else to the same effect) after each review and code modification. That's not necessary. Sometimes I may ask for squash at the end of the review cycle, especially in cases of "history pollution" (unnecessary merges of an upstream branch to the local branch and similar) or excessive trial-and-error commit sequence. But usually even this "final squash" is not necessary.

**Pull request merged**